### PR TITLE
Added integration test for inbound and outbound with any resolver

### PR DESCRIPTION
**Pull request by gandhikrishna**
This PR aims to add integration for inbound and outbound with anyresolver. The test simulates (un)marshalling with anytypes provided no resolver and resolver.

### CODE REVIEW

File: encoding/protobuf/v2/outbound_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+type anyResolver struct {
+	NewMessage proto.Message
+}
+
+func (r anyResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return r.FindMessageByURL(string(message))
+}
+
+func (r anyResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	if r.NewMessage == nil {
```
Comment (jronak) [inline encoding/protobuf/v2/outbound_test.go:119]: We should also assert that the param URL matches expected message url

File: encoding/protobuf/v2/outbound_test.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			))
+
+			client := v2.NewClient(v2.ClientParams{
+				ClientConfig: &transport.OutboundConfig{
+					Outbounds: transport.Outbounds{
+						Unary: out,
+					},
+				},
+				AnyResolver: tt.resolver,
+				Options:     []v2.ClientOption{v2.UseJSON},
```
Comment (jronak) [inline encoding/protobuf/v2/outbound_test.go:89]: Can we test proto and JSON in two different tests?

File: encoding/protobuf/v2/marshal.go [hunk 3]
```diff
@@ -64,7 +64,7 @@ type codec struct {
 func newCodec(anyResolver AnyResolver) *codec {
 	return &codec{
 		jsonMarshaler:   &protojson.MarshalOptions{Resolver: anyResolver},
-		jsonUnmarshaler: &protojson.UnmarshalOptions{Resolver: anyResolver},
+		jsonUnmarshaler: &protojson.UnmarshalOptions{Resolver: anyResolver, DiscardUnknown: true},
```
[the remark this answers is not in the document]
Comment (gandhikrishna) [reply encoding/protobuf/v2/marshal.go:67]: Discardunknown if set to false, while unmarshalling a message with unrecognized fields, preserves the tags and values of such fields in the Message. 
Golang protobuf v2  JSON-unmarshaller tries to unmarshall these unknown fields.  Esp with AnyType, it gives the error `unknown field "@type"` from: https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protojson/decode.go#L156 (that's not the case with gogo JSON-unmarshaller)
Hence setting Discardunknown as true enables json-marshaller not to process unknown/embed fields.

[the remark this answers is not in the document]
Comment (gandhikrishna) [reply encoding/protobuf/v2/marshal.go:67]: This is just with protojson-unmarshaller not with proto-unmarshaller.

File: encoding/protobuf/v2/outbound_test.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+func (r testAnyResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return r.FindMessageByURL(string(message))
+}
+
+func (r testAnyResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	// Custom resolver for TestMessage resolve with both global registered or custom URL
+	if r.NewMessage != nil {
+		if url == "uber.yarpc.encoding.protobuf.TestMessage" || url == "foo.bar.baz" {
```
Comment (jronak) [inline encoding/protobuf/v2/outbound_test.go:182]: nit: we can take exact url from the test case and store it in testAnyResolver struct

**Review by jronak — APPROVED** [review]
lgtm, left a nit

**Pull request merged**